Add common guidance on recording errors on spans and metrics, clarify DB conventions (#1716)
open-telemetry/semantic-conventions: docs/exceptions/README.md

# Semantic Conventions for Exceptions

**Status**: [Stable][DocumentStatus], Unless otherwise specified.

Semantic conventions for Exceptions are defined for the following signals:

* [Exceptions on spans](exceptions-spans.md): Semantic Conventions for Exceptions associated with *spans*.
* [Exceptions in logs](exceptions-logs.md): Semantic Conventions for Exceptions recorded in *logs*.

## Reporting errors in instrumentation code

**Status**: [Development][DocumentStatus]

When instrumented operation fails with an exception, instrumentation SHOULD record
this exception as a [span event](exceptions-spans.md) or a [log record](exceptions-logs.md).

> It's NOT RECOMMENDED to record exceptions that are handled by the instrumented library.

Reviewer: Maybe it is a good idea to include an example of such cases, to avoid any ambiguity/different in interpretations.

# Semantic Conventions for Exceptions

**Status**: [Stable][DocumentStatus], Unless otherwise specified.

Semantic conventions for Exceptions are defined for the following signals:

* [Exceptions on spans](exceptions-spans.md): Semantic Conventions for Exceptions associated with *spans*.
* [Exceptions in logs](exceptions-logs.md): Semantic Conventions for Exceptions recorded in *logs*.

## Reporting exceptions in instrumentation code

**Status**: [Development][DocumentStatus]

> When instrumented operation fails with an exception, instrumentation SHOULD record

Reviewer 2: ```suggestion
When an instrumented operation fails with an exception, instrumentation SHOULD record
```

this exception as a [span event](exceptions-spans.md) or a [log record](exceptions-logs.md).

Recording exceptions on spans SHOULD be accompanied by
- setting span status to `ERROR`
- setting [`error.type`](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/attributes-registry/error.md#error-type)

Refer to the [Recording errors](/docs/general/recording-errors.md) document for additional
details on how to record errors across different signals.

> It's RECOMMENDED to use `Span.recordException` API or logging library API that takes exception instance

Reviewer 2: ```suggestion
It's RECOMMENDED to use the `Span.recordException` API or logging library API that takes exception instance
```

instead of providing individual attributes. This enables the OpenTelemetry SDK to
control what information is recorded based on application configuration.

It's NOT RECOMMENDED to record the same exception more than once.
It's NOT RECOMMENDED to record exceptions that are handled by the instrumented library.

> For example, in this code-snippet, `ResourceNotFoundException` is handled and corresponding

Reviewer 2: ```suggestion
For example, in this code-snippet, `ResourceAlreadyExistsException` is handled and the corresponding
```

> native instrumentation should not record it. Other exceptions, that are propagated
> to the caller, should be recorded (or logged) once.

Reviewer 2: ```suggestion
native instrumentation should not record it. Exceptions which are propagated
to the caller should be recorded (or logged) once.
```

# Semantic Conventions for Exceptions

**Status**: [Stable][DocumentStatus], Unless otherwise specified.

Semantic conventions for Exceptions are defined for the following signals:

* [Exceptions on spans](exceptions-spans.md): Semantic Conventions for Exceptions associated with *spans*.
* [Exceptions in logs](exceptions-logs.md): Semantic Conventions for Exceptions recorded in *logs*.

> ## Reporting exceptions in instrumentation code

Reviewer: Shouldn't this new section be part of the new document `recording-errors.md`? It seems more fitting there than here as part of the general readme.

Also then there's some duplicated information here and in the `recording-errors.md` such as what Span status to set, what to set for error.type and etc.

Author: good point, moved it
